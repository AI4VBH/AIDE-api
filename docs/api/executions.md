# Executions

## `GET /executions/:workflow_instance_id/tasks/:execution_id/artifacts`

This returns the task output artifacts.

### Parameters

| Parameter Type | Label | Description |
|----------------|-------|-------------|
| URL | `:workflow_instance_id` | The workflow instance ID |
| URL | `:execution_id` | The task execution ID |

### Responses

This endpoint returns a dictionary of filename and MINIO object key.

| Code | Description |
|------|-------------|
| `200` | Returns output artifacts |
| `404` | Workflow instance not found or task execution not found. |
| `500` | Server error. The response will be a [Problem details](https://datatracker.ietf.org/doc/html/rfc7807) object with server error details. |

### Example Request

```bash
curl --location --request GET 'http://localhost:5000/executions/9df2f059-7669-4404-b02c-427510d7db04/tasks/c2fa42a0-4a96-46b0-b599-8467ca6d3f7e/artifacts'
```

### Example Response

```json
{
  "dicom-report": "9df2f059-7669-4404-b02c-427510d7db04/c2fa42a0-4a96-46b0-b599-8467ca6d3f7e/dicom-report"
}
```

---

## `GET /executions/artifact-download?key=:object_key`

This redirects the request to a pre-signed URL from MINIO enabling the downloading of file.

### Parameters

| Parameter Type | Label | Description |
|----------------|-------|-------------|
| Query | `:object_key` | The object key for the artifact you are attempting to download |

### Responses

| Code | Description |
|------|-------------|
| `302` | Redirect to the pre-signed MINIO url |
| `400` | Missing `:object_key` |
| `500` | Internal server error |

### Example Request

```bash
curl --location --request GET 'http://localhost:5000/executions/artifact-download?key=9df2f059-7669-4404-b02c-427510d7db04/c2fa42a0-4a96-46b0-b599-8467ca6d3f7e/dicom-report'
```

---

## `GET /executions/:workflow_instance_id/tasks/:execution_id/metadata`

Returns the task execution metadata.

### Response

Returns a dictionary.

| Code | Description |
|------|-------------|
| `200` | Returns execution metadata |
| `404` | Workflow instance not found or task execution not found. |
| `500` | Server error. The response will be a [Problem details](https://datatracker.ietf.org/doc/html/rfc7807) object with server error details. |

### Example Request

```bash
curl --location --request GET 'http://localhost:5000/executions/9df2f059-7669-4404-b02c-427510d7db04/tasks/c2fa42a0-4a96-46b0-b599-8467ca6d3f7e/metadata'
```

### Example Response

```json
{
  "metadata-1": "value",
  "metadata-2": "value"
}
```

