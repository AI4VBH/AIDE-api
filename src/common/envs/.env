PORT=5000

KEYCLOAK_URL='http://0.0.0.0:8443'
KEYCLOAK_MAIN_REALM='master'
KEYCLOAK_REALM='Demo-Realm'
KEYCLOAK_CLIENTID='nest-app'
KEYCLOAK_SECRET='83790b4f-48cd-4b6c-ac60-451a918be4b9'
KEYCLOAK_STATIC_ROLES='["admin", "clinician"]'

DATABASE_HOST='localhost'
DATABASE_NAME='nest_api'
DATABASE_USER='postgres'
DATABASE_PASSWORD='admin'
DATABASE_PORT=5432

ELASTIC_HOST=''
ELASTIC_PORT=''
ELASTIC_USE_SSL=''
ELASTIC_USERNAME=''
ELASTIC_PASSWORD=''
ELASTIC_INDEX=''

MONAI_API_HOST='https://localhost:8081/'
MIG_API_HOST='http://localhost:5010/'

MINIO_HOST='localhost'
MINIO_PORT=9001
MINIO_USE_SSL=false
MINIO_BUCKET='monaideploy'
MINIO_ACCESS_KEY='wOzmbchico52AMp3'
MINIO_SECRET_KEY='zyoT9SiMLiTCvNT0DS636FMiN4yi2258'
