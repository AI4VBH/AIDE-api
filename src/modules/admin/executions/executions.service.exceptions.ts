export enum ExecutionsServiceExceptionType {
  MISSING_TASK = 1,
}

export class ExecutionsServiceException extends Error {
  constructor(type: ExecutionsServiceExceptionType, message: string) {
    super(message);

    this.type = type;
  }

  readonly type: ExecutionsServiceExceptionType;
}
