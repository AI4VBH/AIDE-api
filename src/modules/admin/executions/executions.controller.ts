import {
  BadRequestException,
  Controller,
  Get,
  Inject,
  Param,
  ParseUUIDPipe,
  Query,
  Res,
  UseFilters,
} from '@nestjs/common';
import { Response } from 'express';
import ExternalServerExceptionFilter from 'shared/http/external-server-exception.filter';
import { ExecutionsService } from './executions.service';

@Controller('executions')
@UseFilters(ExternalServerExceptionFilter)
export class ExecutionsController {
  @Inject(ExecutionsService)
  private readonly executionsService: ExecutionsService;

  @Get(':workflow_instance_id/tasks/:execution_id/artifacts')
  async getWorkflowInstanceArtifacts(
    @Param('workflow_instance_id') workflowInstanceId: string,
    @Param('execution_id', ParseUUIDPipe) executionId: string,
  ) {
    return await this.executionsService.getWorkflowInstanceArtifacts(
      workflowInstanceId,
      executionId,
    );
  }

  @Get('artifact-download')
  async getArtifactDownloadUrl(
    @Query('key') file: string,
    @Res() response: Response,
  ) {
    if (!file || !file.trim()) {
      throw new BadRequestException('key query value is missing');
    }

    const { contentType, stream } = await this.executionsService.getArtifact(
      file,
    );

    if (contentType) {
      response.contentType(contentType);
    }

    const piped = stream.pipe(response);

    return piped;
  }

  @Get(':workflow_instance_id/tasks/:execution_id/metadata')
  getWorkflowInstanceMetadata(
    @Param('workflow_instance_id') workflowInstanceId: string,
    @Param('execution_id', ParseUUIDPipe) executionId: string,
  ) {
    return this.executionsService.getWorkflowInstanceMetadata(
      workflowInstanceId,
      executionId,
    );
  }
}
