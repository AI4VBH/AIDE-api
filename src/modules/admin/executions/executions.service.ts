import { HttpService } from '@nestjs/axios';
import { Inject, Injectable } from '@nestjs/common';
import { lastValueFrom } from 'rxjs';
import { MinioClient, MinoiClientException } from 'shared/minio/minio-client';
import internal from 'stream';
import { MonaiWorkflowInstance } from '../workflowinstances/workflowinstances.interface';
import {
  ExecutionsServiceException,
  ExecutionsServiceExceptionType,
} from './executions.service.exceptions';

@Injectable()
export class ExecutionsService {
  @Inject(HttpService)
  private readonly httpService: HttpService;

  @Inject(MinioClient)
  private readonly minioClient: MinioClient;

  public get minioBaseUrl(): string {
    return this.minioClient.baseUrl;
  }

  async getWorkflowInstanceArtifacts(
    workflowInstanceId: string,
    executionId: string,
  ) {
    const response = await lastValueFrom(
      this.httpService.get<MonaiWorkflowInstance>(
        `/workflowinstances/${workflowInstanceId}`,
      ),
    );

    const task = (response.data?.tasks ?? []).find(
      (t) => t.execution_id === executionId,
    );

    if (!task) {
      throw new ExecutionsServiceException(
        ExecutionsServiceExceptionType.MISSING_TASK,
        'Task not found',
      );
    }

    const artifacts = task?.output_artifacts ?? {};

    return artifacts;
  }

  async getArtifactUrl(file: string) {
    try {
      const url = await this.minioClient.getPresignedObjectUrl(file);
      return url;
    } catch (exception) {
      throw new MinoiClientException(exception.message);
    }
  }

  async getArtifact(
    file: string,
  ): Promise<{ contentType: string; stream: internal.Readable }> {
    try {
      const stream = await this.minioClient.getObjectByName(file);
      const metadata = await this.minioClient.getObjectMetadata(file);

      return { stream, contentType: metadata['content-type'] };
    } catch (exception) {
      throw new MinoiClientException(exception.code);
    }
  }

  async getWorkflowInstanceMetadata(
    workflowInstanceId: string,
    executionId: string,
  ) {
    const response = await lastValueFrom(
      this.httpService.get<MonaiWorkflowInstance>(
        `/workflowinstances/${workflowInstanceId}`,
      ),
    );

    const task = (response.data?.tasks ?? []).find(
      (t) => t.execution_id === executionId,
    );

    if (!task) {
      throw new ExecutionsServiceException(
        ExecutionsServiceExceptionType.MISSING_TASK,
        'Task not found',
      );
    }

    return task?.result ?? {};
  }
}
