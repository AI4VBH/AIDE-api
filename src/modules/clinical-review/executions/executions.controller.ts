import { Controller, Get } from '@nestjs/common';
import { Executions, ExecutionsSearchResult } from './executions.interface';
import { ExecutionsService } from './executions.service';

@Controller('executions')
export class ExecutionsController {
  constructor(private readonly executionsService: ExecutionsService) {}

  @Get()
  async getModels() {
    return await this.executionsService.getExecutions();
  }
}
