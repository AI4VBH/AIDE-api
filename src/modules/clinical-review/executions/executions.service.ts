import { Injectable } from '@nestjs/common';
import { Executions, ExecutionsSearchResult } from './executions.interface';

@Injectable()
export class ExecutionsService {
  async getExecutions() {
    // console.log('searching');
    // console.log(searchResult);
    return {};
  }
}
