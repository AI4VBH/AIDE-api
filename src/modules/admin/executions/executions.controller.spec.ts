/*
 * Copyright 2022 Guy’s and St Thomas’ NHS Foundation Trust
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { createMock, DeepMocked } from '@golevelup/ts-jest';
import { BadRequestException, Logger } from '@nestjs/common';
import { ConfigModule } from '@nestjs/config';
import { Test, TestingModule } from '@nestjs/testing';
import { Readable } from 'stream';
import { ExecutionsController } from './executions.controller';
import { ExecutionsService } from './executions.service';
import { Response } from 'express';

describe('ExecutionsController', () => {
  let controller: ExecutionsController;
  let executionsService: DeepMocked<ExecutionsService>;

  beforeEach(async () => {
    executionsService = createMock<ExecutionsService>();

    const module: TestingModule = await Test.createTestingModule({
      imports: [
        ConfigModule.forRoot({
          isGlobal: true,
          load: [() => ({})],
        }),
      ],
      controllers: [ExecutionsController],
      providers: [
        { provide: ExecutionsService, useValue: executionsService },
        {
          provide: Logger,
          useFactory: () => createMock<Logger>(),
        },
      ],
    }).compile();

    controller = module.get<ExecutionsController>(ExecutionsController);
  });

  it('should be defined', () => {
    expect(controller).toBeDefined();
  });

  describe('getWorkflowInstanceArtifacts', () => {
    it('returns expected result', async () => {
      executionsService.getWorkflowInstanceArtifacts.mockResolvedValue({
        'file.ext': 'minio-object-key/file.ext',
        'file-2.ext': 'minio-object-key/file-2.ext',
      });

      const result = await controller.getWorkflowInstanceArtifacts(
        'c7ec91ef-7228-441c-9f74-a5d09384ccb6',
        '5256047b-ac8b-41eb-b8cb-072678992081',
      );

      expect(result).toMatchSnapshot();
    });
  });

  describe('getWorkflowInstanceMetadata', () => {
    it('returns expected result', async () => {
      executionsService.getWorkflowInstanceMetadata.mockResolvedValue({
        'stat-1': 'value',
        'stat-2': 'value 2',
      });

      const result = await controller.getWorkflowInstanceMetadata(
        '889ea3b6-9123-4a1a-81c1-6c84e9152783',
        '0c195a19-2631-4f28-9654-0b1ed6e6925d',
      );

      expect(result).toMatchSnapshot();
    });
  });

  describe('getArtifactDownloadUrl', () => {
    it('returns expected result', async () => {
      executionsService.getArtifacts.mockResolvedValue([
        {
          name: 'file-1.dcm.json',
          stream: createMock<Readable>(),
        },
        {
          name: 'file-1.dcm',
          stream: createMock<Readable>(),
        },
      ]);

      const response = createMock<Response>();

      await controller.getArtifactDownloadUrl('file.ext', response);

      expect(executionsService.getArtifacts).toHaveBeenCalled();
    });

    it.each(['', ' ', null, undefined])(
      'throws exception when key is %s',
      async (key: string) => {
        const response = createMock<Response>();

        const action = async () =>
          await controller.getArtifactDownloadUrl(key, response);

        await expect(action).rejects.toThrow(BadRequestException);
      },
    );
  });
});
