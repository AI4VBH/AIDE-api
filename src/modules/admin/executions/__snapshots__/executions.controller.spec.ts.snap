// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExecutionsController getWorkflowInstanceArtifacts returns expected result 1`] = `
{
  "file-2.ext": "minio-object-key/file-2.ext",
  "file.ext": "minio-object-key/file.ext",
}
`;

exports[`ExecutionsController getWorkflowInstanceMetadata returns expected result 1`] = `
{
  "stat-1": "value",
  "stat-2": "value 2",
}
`;
